File: backend/app/schemas/retention.py
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field
from datetime import datetime, date

from app.schemas.analytics import StandardAnalyticsResponse, KPIResponse

# -----------------------------------------
# Retention Cohort Matrix
# -----------------------------------------

class CohortCell(BaseModel):
    period: int = Field(..., description="The period offset (e.g., Month 1, Month 2)")
    active_users: int = Field(..., description="Number of users still active in this period")
    retention_percentage: float = Field(..., description="Percentage of original cohort retained")

class CohortRow(BaseModel):
    cohort_date: str = Field(..., description="The starting date identifier for the cohort (e.g., '2026-01')")
    initial_size: int = Field(..., description="Total users in the initial cohort")
    cells: List[CohortCell] = Field(default_factory=list)

class CohortMatrixResponse(BaseModel):
    granularity: str = Field(..., description="'daily', 'weekly', 'monthly', 'quarterly'")
    rows: List[CohortRow] = Field(default_factory=list)

# -----------------------------------------
# Retention Health & Personas
# -----------------------------------------

class UserPersonaResponse(BaseModel):
    persona_name: str = Field(..., description="e.g., 'Power User', 'Dormant'")
    user_count: int
    percentage_of_base: float
    description: Optional[str] = None

class RetentionHealthResponse(BaseModel):
    overall_health_score: float = Field(..., description="0-100 normalized score")
    health_category: str = Field(..., description="Excellent, Good, Average, Poor, Critical")
    personas: List[UserPersonaResponse] = Field(default_factory=list)

# -----------------------------------------
# Churn Intelligence
# -----------------------------------------

class ChurnSegmentResponse(BaseModel):
    segment_name: str = Field(..., description="e.g., 'Voluntary Cancel', 'Payment Failed'")
    churn_count: int
    churn_rate: float

class ChurnDistributionResponse(BaseModel):
    total_churned: int
    overall_churn_rate: float
    segments: List[ChurnSegmentResponse] = Field(default_factory=list)

# -----------------------------------------
# Retention Trends
# -----------------------------------------

class RetentionTrendPoint(BaseModel):
    date: str
    retained_users: int
    churned_users: int
    retention_rate: float

class RetentionTrendResponse(BaseModel):
    metric_name: str
    granularity: str
    data_points: List[RetentionTrendPoint] = Field(default_factory=list)

# -----------------------------------------
# Customer Lifecycle
# -----------------------------------------

class LifecycleTransition(BaseModel):
    source_state: str
    target_state: str
    user_count: int

class CustomerLifecycleResponse(BaseModel):
    period_start: date
    period_end: date
    transitions: List[LifecycleTransition] = Field(default_factory=list)

File: frontend/src/api/retention.ts
import axios from 'axios';

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000/api/v1';

export interface StandardAnalyticsResponse<T> {
  data: T;
  filters_applied: Record<string, any>;
  timestamp: string;
}

export interface KPIMetric {
  name: string;
  value: number;
  delta_percentage: number;
  trend_direction: 'up' | 'down' | 'flat';
}

export interface KPICollectionResponse {
  metrics: KPIMetric[];
}

export interface CohortCell {
  period: number;
  active_users: number;
  retention_percentage: number;
}

export interface CohortRow {
  cohort_date: string;
  initial_size: number;
  cells: CohortCell[];
}

export interface CohortMatrixResponse {
  granularity: string;
  rows: CohortRow[];
}

export interface UserPersona {
  persona_name: string;
  user_count: number;
  percentage_of_base: number;
  description?: string;
}

export interface SubScoreBreakdown {
  activity_score: number;
  volume_score: number;
  session_quality_score: number;
  feature_breadth_score: number;
}

export interface RetentionHealthResponse {
  overall_health_score: number;
  health_category: 'Excellent' | 'Good' | 'Average' | 'Poor' | 'Critical';
  personas: UserPersona[];
}

export interface ChurnSegment {
  segment_name: string;
  churn_count: number;
  churn_rate: number;
}

export interface ChurnDistributionResponse {
  total_churned: number;
  overall_churn_rate: number;
  segments: ChurnSegment[];
}

export interface TrendPoint {
  date: string;
  value: number;
}

export interface TrendSeriesResponse {
  metric_name: string;
  data_points: TrendPoint[];
}

export interface ExecutiveInsight {
  title: string;
  category: 'Positive' | 'Negative' | 'Actionable';
  impact: 'High' | 'Medium' | 'Low';
  description: string;
}

// -----------------------------------------
// API Fetchers
// -----------------------------------------

export const fetchRetentionOverview = async (params?: Record<string, any>): Promise<StandardAnalyticsResponse<any>> => {
  const res = await axios.get(`${API_BASE_URL}/analytics/overview`, { params });
  return res.data;
};

export const fetchChurnKPIs = async (params?: Record<string, any>): Promise<StandardAnalyticsResponse<KPICollectionResponse>> => {
  const res = await axios.get(`${API_BASE_URL}/retention/churn/kpis`, { params });
  return res.data;
};

export const fetchRetentionHealth = async (params?: Record<string, any>): Promise<StandardAnalyticsResponse<RetentionHealthResponse>> => {
  const res = await axios.get(`${API_BASE_URL}/retention/health`, { params });
  return res.data;
};

export const fetchCohortMatrix = async (granularity = 'monthly', params?: Record<string, any>): Promise<StandardAnalyticsResponse<CohortMatrixResponse>> => {
  const res = await axios.get(`${API_BASE_URL}/retention/cohorts`, { params: { granularity, ...params } });
  return res.data;
};

export const fetchChurnDistribution = async (params?: Record<string, any>): Promise<StandardAnalyticsResponse<ChurnDistributionResponse>> => {
  const res = await axios.get(`${API_BASE_URL}/retention/churn`, { params });
  return res.data;
};

export const fetchRetentionTrends = async (params?: Record<string, any>): Promise<StandardAnalyticsResponse<TrendSeriesResponse>> => {
  const res = await axios.get(`${API_BASE_URL}/retention/trends`, { params });
  return res.data;
};

export const fetchExecutiveInsights = async (params?: Record<string, any>): Promise<StandardAnalyticsResponse<ExecutiveInsight[]>> => {
  const res = await axios.get(`${API_BASE_URL}/retention/insights`, { params });
  return res.data;
};
